File: Sources/NockerlDesign/Components/NockerlAvatar.swift
// NockerlAvatar: person/session identity as a clipped disc (the compose/react
// canon, avatar.mdx). A strict image → initials fallback ladder, the ratified
// size ramp, and an optional presence dot notched into the host surface.
// Replaces the interim initials disc the Swift AgentWidget shipped with.

import SwiftUI

/// The ratified avatar diameter ramp (avatar.mdx).
public enum NockerlAvatarSize {
    /// 24pt: dense rows, stacks.
    case xs
    /// 32pt: chrome (top bar).
    case sm
    /// 40pt: sheets and headers.
    case md
    /// 48pt: profile blocks.
    case lg
    /// 64pt: hero identity.
    case xl

    var diameter: CGFloat {
        switch self {
        case .xs: return 24
        case .sm: return 32
        case .md: return 40
        case .lg: return 48
        case .xl: return 64
        }
    }

    var initialsSize: CGFloat {
        switch self {
        case .xs: return NockerlFontSize.size10
        case .sm: return NockerlFontSize.size12
        case .md: return NockerlFontSize.size14
        case .lg: return NockerlFontSize.size16
        case .xl: return NockerlFontSize.size24
        }
    }
}

/// The identity disc: image → initials on the soft accent wash; optional
/// bottom-right presence dot notched by a host-surface ring. The dot is
/// decorative. Fold the state into the accessibility label (law §13).
public struct NockerlAvatar: View {
    private let name: String
    private let image: Image?
    private let size: NockerlAvatarSize
    private let presence: Color?
    private let presenceRing: Color?
    private let accessibilityName: String?

    @Environment(\.colorScheme) private var colorScheme

    /// Create an avatar.
    /// - Parameters:
    ///   - name: the person/session name (drives initials + the a11y default).
    ///   - image: optional avatar image; initials render when `nil`.
    ///   - size: the ratified ramp (default `.sm`, 32pt).
    ///   - presence: optional presence-dot color (a `dot*` palette token).
    ///   - presenceRing: the notch ring; match the HOST surface (defaults to
    ///     `cardSurface1`).
    ///   - accessibilityName: override the accessible name (fold presence in).
    public init(
        name: String,
        image: Image? = nil,
        size: NockerlAvatarSize = .sm,
        presence: Color? = nil,
        presenceRing: Color? = nil,
        accessibilityName: String? = nil
    ) {
        self.name = name
        self.image = image
        self.size = size
        self.presence = presence
        self.presenceRing = presenceRing
        self.accessibilityName = accessibilityName
    }

    public var body: some View {
        let palette = NockerlPalette.resolve(colorScheme)

        ZStack(alignment: .bottomTrailing) {
            Group {
                if let image {
                    image
                        .resizable()
                        .scaledToFill()
                } else {
                    ZStack {
                        palette.accentPrimarySoft
                        Text(Self.initials(from: name))
                            // 500 is the bold cap (law §11): medium, never bold.
                            .font(.nockerl(size: size.initialsSize, weight: .medium))
                            .foregroundColor(palette.accentPrimary)
                            .lineLimit(1)
                    }
                }
            }
            .frame(width: size.diameter, height: size.diameter)
            .clipShape(Circle())

            if let presence {
                ZStack {
                    Circle()
                        .fill(presenceRing ?? palette.cardSurface1)
                        .frame(width: 12, height: 12)
                    Circle()
                        .fill(presence)
                        .frame(width: 8, height: 8)
                }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityName ?? name)
    }

    /// "Ada Lovelace" → "AL"; blank → "?".
    static func initials(from name: String) -> String {
        let parts = name
            .split(separator: " ")
            .prefix(2)
            .compactMap { $0.first.map(String.init) }
        let joined = parts.joined().uppercased()
        return joined.isEmpty ? "?" : joined
    }
}

File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlAvatar.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors

/**
 * The standardized **avatar**: person/session identity as a clipped disc with a
 * strict fallback ladder (avatar.mdx): image → initials. Unifies the raw
 * `Image(...).clip(CircleShape)` pattern client #1 re-derives (32dp in the top
 * bar, 40dp in the settings sheet) into one wrapper with the ratified size ramp.
 *
 * - **Image**: pass a [painter] (the library takes a [Painter], never an image
 *   loader dependency. Hand it `painterResource(...)` or a Coil painter).
 * - **Initials**: derived from [name] ("Ada Lovelace" → "PM") on the soft
 *   accent wash; a blank name renders the "?" placeholder glyph.
 * - **Presence**: an optional bottom-right dot ([presence], pass a palette dot
 *   token), notched into the host surface via a [presenceRing] ring so the dot
 *   reads anchored, not floating. The dot is decorative here. Fold the state
 *   into [contentDescription] ("the design lead, active"), never color alone (law §13).
 *
 * @param name the person/session name (drives initials + the default a11y name).
 * @param modifier outer modifier.
 * @param painter optional avatar image; initials render when `null`.
 * @param size the ratified diameter ramp (default [NockerlAvatarSize.SM], 32dp).
 * @param shape [CircleShape] (default): the rounded-square variant is a proposed
 *   alternative; pass the control shape once that ratifies.
 * @param presence optional presence-dot color (a `dot*` palette token).
 * @param presenceRing the notch ring color: match the HOST surface the avatar
 *   sits on (defaults to `cardSurface1`).
 * @param contentDescription accessible name (defaults to [name]).
 */
@Composable
fun NockerlAvatar(
    name: String,
    modifier: Modifier = Modifier,
    painter: Painter? = null,
    size: NockerlAvatarSize = NockerlAvatarSize.SM,
    shape: Shape = CircleShape,
    presence: Color? = null,
    presenceRing: Color = LocalNockerlColors.current.cardSurface1,
    contentDescription: String? = null,
) {
    val colors = LocalNockerlColors.current
    // Distinct local: the semantics receiver's property shadows the parameter.
    val accessibleName = contentDescription ?: name

    Box(modifier = modifier.semantics { this.contentDescription = accessibleName }) {
        Box(
            modifier = Modifier.size(size.diameter).clip(shape),
            contentAlignment = Alignment.Center,
        ) {
            if (painter != null) {
                Image(
                    painter = painter,
                    contentDescription = null, // named by the outer semantics
                    modifier = Modifier.fillMaxSize(),
                    contentScale = ContentScale.Crop,
                )
            } else {
                Box(
                    modifier = Modifier.fillMaxSize().background(colors.accentPrimarySoft),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = nockerlAvatarInitials(name),
                        style = size.initialsStyle(),
                        // 500 is the bold cap (law §11). Initials are medium, never bold.
                        fontWeight = FontWeight.Medium,
                        color = colors.accentPrimary,
                        maxLines = 1,
                    )
                }
            }
        }

        if (presence != null) {
            // Bottom-right presence dot, notched by a host-surface ring.
            Box(
                modifier =
                    Modifier
                        .align(Alignment.BottomEnd)
                        .size(NockerlStatusDotDefaults.LoudSize + PRESENCE_RING_WIDTH * 2)
                        .clip(CircleShape)
                        .background(presenceRing),
                contentAlignment = Alignment.Center,
            ) {
                Box(
                    modifier =
                        Modifier
                            .size(NockerlStatusDotDefaults.LoudSize)
                            .clip(CircleShape)
                            .background(presence),
                )
            }
        }
    }
}

/** The ratified avatar diameter ramp (avatar.mdx). */
enum class NockerlAvatarSize(
    /** The disc diameter. */
    val diameter: Dp,
) {
    /** 24dp: dense rows, stacks. */
    XS(24.dp),

    /** 32dp: chrome (client #1's top bar). */
    SM(32.dp),

    /** 40dp: sheets and headers (client #1's settings sheet). */
    MD(40.dp),

    /** 48dp: profile blocks. */
    LG(48.dp),

    /** 64dp: hero identity. */
    XL(64.dp),
    ;

    /** The initials type role scaled to the disc. */
    @Composable
    fun initialsStyle() =
        when (this) {
            XS, SM -> MaterialTheme.typography.labelMedium
            MD, LG -> MaterialTheme.typography.titleSmall
            XL -> MaterialTheme.typography.titleLarge
        }
}

/**
 * Derive avatar initials: first letter of the first + last whitespace-separated
 * tokens, uppercased ("Ada Lovelace" → "PM"; "Nockerl" → "N"); a blank name
 * falls back to the "?" placeholder glyph.
 */
internal fun nockerlAvatarInitials(name: String): String {
    val tokens = name.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
    return when {
        tokens.isEmpty() -> "?"
        tokens.size == 1 -> tokens.first().take(1).uppercase()
        else -> (tokens.first().take(1) + tokens.last().take(1)).uppercase()
    }
}

/** Presence-ring thickness: the notch that anchors the dot into the host surface. */
private val PRESENCE_RING_WIDTH = 2.dp
